sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala

```diff
@@ -204,11 +208,55 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
   private def build(segments: String*): String =
     segments.map(_.trim).filter(_.nonEmpty).mkString(" ")
 
+  private def updateQualifier(
+      expr: Expression,
+      qualifiers: Seq[(String, AttributeSet)]): Expression = {
+    if (qualifiers.isEmpty) {
+      expr
+    } else {
+      expr transform {
+        case a: Attribute =>
+          val index = qualifiers.indexWhere {
+            case (_, inputAttributes) => inputAttributes.contains(a)
+          }
+          if (index == -1) {
+            a
+          } else {
+            a.withQualifiers(qualifiers(index)._1 :: Nil)
+          }
+      }
+    }
+  }
+
+  private def findQualifiers(input: LogicalPlan): Seq[(String, AttributeSet)] = {
+    val results = mutable.ArrayBuffer.empty[(String, AttributeSet)]
+    val nodes = mutable.Stack(input)
+
+    while (nodes.nonEmpty) {
+      val node = nodes.pop()
+      node match {
+        case SubqueryAlias(alias, child) => results += alias -> child.outputSet
+        case _ => node.children.foreach(nodes.push)
+      }
+    }
+
+    results.toSeq
+  }
```

Comment:
So this method is basically a DFS search for all the outermost `SubqueryAlias` operators. Maybe the following version is clearer:

```scala
def findOutermostQualifiers(input: LogicalPlan): Seq[(String, AttributeSet)] = {
  input.collectFirst {
    case SubqueryAlias(alias, child) => Seq(alias -> child.outputSet)
    case plan => plan.children.flatMap(findOutermostQualifiers)
  }.toSeq.flatten
}
```